dhalli/Main.hs

```diff
@@ -0,0 +1,235 @@
+{-# language FlexibleContexts #-}
+{-# language NamedFieldPuns #-}
+{-# language OverloadedStrings #-}
+
+module Main ( main ) where 
+
+import Control.Exception ( SomeException(SomeException), displayException, throwIO )
+import Control.Monad.State.Strict ( evalStateT )
+import Control.Monad.State.Class ( MonadState, get, modify )
+import Control.Monad.IO.Class ( MonadIO, liftIO )
+
+import qualified Data.Map.Strict as StrictMap
+import qualified Data.Text.Lazy as LazyText 
+import qualified Data.Text.Prettyprint.Doc as Pretty
+import qualified Data.Text.Prettyprint.Doc.Render.Terminal as Pretty ( renderIO )
+import qualified Dhall
+import qualified Dhall.Context
+import qualified Dhall.Core as Dhall ( Var(V), Expr, normalize, subst )
+import qualified Dhall.Pretty
+import qualified Dhall.Core as Expr ( Expr(..) )
+import qualified Dhall.Import as Dhall
+import qualified Dhall.Parser as Dhall
+import qualified Dhall.TypeCheck as Dhall
+import qualified System.Console.Haskeline.MonadException as Haskeline
+import qualified System.Console.Repline as Repline
+import qualified System.IO
+import qualified Text.Trifecta.Delta as Trifecta
+
+
+main :: IO ()
+main =
+  evalStateT 
+    ( Repline.evalRepl
+        "⊢ "
+        ( dontCrash . eval )
+        options
+        ( Repline.Word completer )
+        greeter
+    )
+    emptyEnv
+
+
+data Env = Env
+  { envBindings :: StrictMap.Map Dhall.Text Binding
```

Comment:
> If you used a Map for that then the second x would override the first x and you would turn y from a well-typed to an ill-typed term.

That's not quite true, as expressions are normalized before putting them into the environment. So in this case, you'd add `(x, 1)`, then `(y, 1)`, then replace `(x, True)`.

You're otherwise right though, so I'll make that change.